### PR TITLE
[SPARK-17192] [SQL] Issue Exception when Users Specify the Partitioning Columns without a Given Schema

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?

Address the comments by @yhuai in the original PR: https://github.com/apache/spark/pull/14207

First, issue an exception instead of logging a warning when users specify the partitioning columns without a given schema. 

Second, refactor the codes a little. 
### How was this patch tested?

Fixed the test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
retest this please

**Comment by yhuai** [comment]
sorry. I missed this PR. Can you update?

**Comment by gatorsmile** [comment]
No problem. The conflicts are resolved. Thanks! @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 1]
```diff
@@ -72,29 +72,19 @@ case class PreprocessDDL(conf: SQLConf) extends Rule[LogicalPlan] {
...
-            s"ignored. The schema and partition columns of table ${tableDesc.identifier} will " +
-            "be inferred.")
-        c.copy(tableDesc = tableDesc.copy(partitionColumnNames = Nil))
-      } else {
-        c
+      if (tableDesc.partitionColumnNames.nonEmpty) {
+        failAnalysis("Cannot specify partition information if the table schema is not specified " +
+          "when creating and will be inferred at runtime")
       }
+      c
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Because of this change, I think we need to create a new jira.

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: A new JIRA is created and the PR title is updated. Thanks!

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: How about `It is not allowed to specify partition columns when the table schema is not defined. When the table schema is not provided, schema and partition columns will be inferred.`?

**Comment by yhuai** [comment]
LGTM. I left one comment about the error message

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Sounds better to me. Thanks! Let me change it now.

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
ping @yhuai : )

**Comment by yhuai** [comment]
LGTM. Thanks. Merging to master.

**Pull request closed without merging**